Replicate docs with attachments

jeqq opened the pull request.

jeqq committed:
Make possible to replicate content with attachments. Work in progress.

jeqq committed:
Check format.

jeqq committed:
Fix access problems when replicating entities with attachments.

jeqq committed:
Fixtures fixes.

jeqq committed:
Fix file attachment denormalization.

jeqq committed:
Fixtures fixes.

jeqq committed:
Remove taxonomy term references for fixtures.

jeqq committed:
Fix creation of the files with the same uri.

jeqq committed:
Fix permissions for replicator role.

jeqq committed:
Revert "Fix broken tests."

This reverts commit b8cb46340083d4df2b4d42dbd883630ebba31783.

jeqq committed:
Show error and access logs.

jeqq committed:
Invalidate 'file_list' cache tags in order to save a new file.

jeqq committed:
Configuration changes.

jeqq committed:
Clear image effect cache in order to save a node with attachments.

jeqq committed:
Use filename for alternative text value.

jeqq committed:
Fixtures and file denormalization fixes.

jeqq committed:
Save file attachments in the rest plugin not in serializer.

jeqq committed:
Work with Drupal 8 dev branch.

dickolsson commented on relaxed.module:
```diff
@@ -73,6 +73,26 @@ function relaxed_entity_access(EntityInterface $entity, $operation, AccountInter
+    if (in_array($entity_bundle, array_keys($bundles))){
+      $entity_type = \Drupal::entityManager()->getDefinition($entity_type_id);
+      if ($multiversion_manager->isSupportedEntityType($entity_type) || $entity_type_id == 'replication_log') {
+        return AccessResult::allowedIfHasPermission($account, 'perform content replication');
```
Why do we need this specifically for attachment replication?

jeqq replied:
@dickolsson This maybe is not specifically for attachment replication, it fixes an access problem I've detected when doing this replication.

dickolsson commented on src/Controller/ResourceController.php:
```diff
@@ -160,6 +164,49 @@ public function handle(Request $request) {
+                  \Drupal::service('plugin.manager.image.effect')->clearCachedDefinitions();
+                  if ($file instanceof FileInterface) {
+                    Cache::invalidateTags(array('file_list'));
+                    $file->save();
```
Instead of saving the file directly here in the controller, I think that we should load the attachment resource plugin (`Drupal\relaxed\Plugin\rest\resource\AttachmentResource`) and pass it to the `AttachmentResource::put()` method on that plugin that will be responsible for saving.

jeqq replied:
@dickolsson I don't know if this is a better way to do this because when using the `AttachmentResource::put()` method we should already have the entity saved. At the moment we firstly denormalize and save the attachments and then denormalize and save the entity (for multipart/related requests).

dickolsson commented on src/Normalizer/ContentEntityNormalizer.php:
```diff
@@ -249,6 +272,11 @@ public function denormalize($data, $class, $format = NULL, array $context = arra
       }
     }
 
+    // Remove changed info, otherwise we can get validation errors.
```
Can you explain this a bit more? Why can we get validation errors?

dickolsson commented on src/Plugin/rest/resource/DbResource.php:
```diff
@@ -125,6 +126,27 @@ public function post($workspace, ContentEntityInterface $entity = NULL) {
+            unset($entity->{$field_name}[$delta]->entity_to_save);
+          }
+        }
+      }
```
Why do we need this logic here? Isn't the replication using the Bulk API endpoint? What part of the process is using POST on the db endpoint?

jeqq replied:
@dickolsson The replication doesn't use this endpoint to POST documents, but we can do this. An example when we use it is when we create the documents in the source database before replication (for testing):

`curl -X POST -H "Content-Type: application/json" -d "$document"  localhost:5984/source`

dickolsson commented on src/RelaxedServiceProvider.php:
```diff
@@ -19,8 +19,10 @@ class RelaxedServiceProvider implements ServiceModifierInterface {
+      // Adds related as known format.
+      $container->getDefinition('http_middleware.negotiation')->addMethodCall('registerFormat', ['related', ['multipart/related']]);
+      // Adds mixed as known format.
       $container->getDefinition('http_middleware.negotiation')->addMethodCall('registerFormat', ['mixed', ['multipart/mixed']]);
```
Let's do something like this instead:

``` php
$negotiation = $container->getDefinition('http_middleware.negotiation');
$negotiation->addMethodCall(...);
$negotiation->addMethodCall(...);
```

dickolsson commented on src/HttpMultipart/Message/MultipartResponse.php:
```diff
@@ -57,10 +57,15 @@ public static function parseMultipartBody(StreamInterface $stream)
+          if (count($header_parts) == 2) {
+            list($key, $value) = $header_parts;
+            $headers[strtolower($key)] = $value;
+          }
```
Let's add some comments to what's going on. It's some pretty complicated logic here :smile:

jeqq committed:
Add comments and code refactoring.

jeqq committed:
Change the branch for multiversion module.

jeqq committed:
Save attachment in resource plugin, not in controller.

jeqq committed:
Test with more than one worker process - 4 (default).

dickolsson commented:
@jeqq What's the next step with this PR, now that #32 is merged?

jeqq committed:
Fix multipart/related request processing.

jeqq committed:
Change the branch for multiversion.

jeqq committed:
Entity reference field replication fixes.

jeqq committed:
Add PHP configuration file with additional configs.

jeqq committed:
Show log files content after couuch-to-drupal replication.

jeqq committed:
Travis configuration changes.

jeqq committed:
Entity references stub entity fixes.

jeqq committed:
Stub entity creation fixes.

jeqq committed:
Replication log entity type fixes.

Add uuid field and make possible to get the replication_log with GET
when doing replication.

jeqq committed:
Return the correct format for replication_log entity type.

jeqq committed:
Make recorded_seq not required - fix for pouchdb replication.

dickolsson commented on .travis.yml:
```diff
@@ -56,17 +59,14 @@ install:
   - git clone --depth 1 --branch 8.0.x http://git.drupal.org/project/drupal.git $TRAVIS_BUILD_DIR/../drupal
   - git clone --branch 8.x-1.x https://github.com/dickolsson/drupal-key_value.git $TRAVIS_BUILD_DIR/../drupal/modules/key_value
-  - git clone --branch 8.x-1.x https://github.com/dickolsson/drupal-multiversion.git $TRAVIS_BUILD_DIR/../drupal/modules/multiversion
+  - git clone --branch new-revision-creation-fixes https://github.com/dickolsson/drupal-multiversion.git $TRAVIS_BUILD_DIR/../drupal/modules/multiversion
```
@jeqq Can you please open a new multiversion PR for this branch so that I can start reviewing it?

jeqq committed:
Create unique names for stub terms.

Commit:
Changing back to correct multiversion branch

Commit:
Adding @todo for better code separation later

jeqq committed:
Try to fix stub entities cretion and updates.

jeqq committed:
Remove unused constant.

jeqq committed:
Disable caching for relaxed routes.

jeqq committed:
Travis configuration fixes.

jeqq committed:
Fix broken travis script after rebase.

jeqq committed:
Revert "updating bash scripts to work locally"

jeqq committed:
Revert "updating couch-to-drupal for local"

This reverts commit 7164bb185a1c18424d6bc11a7dc9c291e23f2734.

jeqq committed:
Configuration fixes.

jeqq committed:
Configuration fixes.

dickolsson commented on .travis.yml:
```diff
@@ -60,15 +63,15 @@ install:
   - mysql -e "SHOW VARIABLES LIKE 'wait_timeout';"
 
+  - composer selfupdate
+  - composer global require drush/drush:dev-master
```
Why do we need to install Composer twice in Travis?

jeqq replied:
@dickolsson It looks like I added it accidentally. I've removed these lines.

jeqq committed:
Remove duplicate code.

timmillwood committed:
re-enabling caching

timmillwood committed:
set MaxAge 0 for all db routes

timmillwood committed:
removing max age

timmillwood committed:
Changing response order

timmillwood committed:
disabling caching with definition

timmillwood committed:
fixing test after changing order

timmillwood committed:
Converting instance_start_time to a string and adding tests for it

timmillwood committed:
merging caching

jeqq committed:
Try to fix permission error on user content type replication.

jeqq committed:
Stub entity processor fixes.

jeqq committed:
Exclude stub entities from changes endpoint results.

jeqq committed:
Remove pass field from fixtures.

timmillwood committed:
adding php-client test

timmillwood committed:
£

timmillwood committed:
Merge branch 'replicate-docs-with-attachments' of github.com:dickolsson/drupal-relaxed into caching

jeqq committed:
Collect bubbleable metadata in a render context for resource responces.

jeqq committed:
Add values for stub creation for comment entity type.

timmillwood committed:
adding composer.lock and calling json object

timmillwood committed:
using array instead of object

timmillwood committed:
Adding no attachments fixtures

timmillwood committed:
Merge branch 'replicate-docs-with-attachments' of github.com:dickolsson/drupal-relaxed into caching

timmillwood committed:
need to add no-attachments-documents to COUNT

jeqq committed:
Cahing and normalization fixes.

jeqq committed:
Fix response and cache processing in controller.
